Remove send on closed channel

In the devnet, the following error was observed after some of the nodes panicked:

```
panic: send on closed channel

goroutine 128 [running]:
github.com/dusk-network/dusk-blockchain/pkg/p2p/peer.sendError(0xc0023f9140, 0x108c640, 0xc01fa7b310)
        /work/pkg/p2p/peer/peer.go:211 +0x49
github.com/dusk-network/dusk-blockchain/pkg/p2p/peer.(*Writer).writeLoop(0xc002231110, 0x10a2c40, 0xc00245ac00, 0xc0023f9080, 0xc0023f9140)
        /work/pkg/p2p/peer/peer.go:295 +0x650
github.com/dusk-network/dusk-blockchain/pkg/p2p/peer.(*Writer).Serve(0xc002231110, 0x10a2c40, 0xc00245ac00, 0xc0023f9080, 0xc0023f9140)
        /work/pkg/p2p/peer/peer.go:225 +0xd8
created by github.com/dusk-network/dusk-blockchain/pkg/p2p/peer.Create
        /work/pkg/p2p/peer/peer.go:202 +0x106
```

This is accompanied by the logs showing that some connections are dropped:

```
{"error":"write tcp 159.89.24.209:58020-\u003e165.227.33.253:7100: write: broken pipe","level":"error","msg":"got error back from keepAlive","process":"keepaliveloop","time":"2021-04-16T14:29:47Z"}
{"coordinator_round":45145,"level":"debug","msg":"storing future round for later","process":"agreement","round":45147,"time":"2021-04-16T14:29:53Z","topic":"Agreement"}
{"error":"write tcp 159.89.24.209:60666-\u003e104.131.173.173:7100: write: broken pipe","level":"error","msg":"got error back from keepAlive","process":"keepaliveloop","time":"2021-04-16T14:29:54Z"}
{"error":"write tcp 159.89.24.209:7100-\u003e143.110.156.197:55554: write: broken pipe","level":"error","msg":"got error back from keepAlive","process":"keepaliveloop","time":"2021-04-16T14:29:55Z"}
{"coordinator_round":45145,"level":"debug","msg":"storing future round for later","process":"agreement","round":45148,"time":"2021-04-16T14:30:11Z","topic":"Agreement"}
{"coordinator_round":45145,"level":"debug","msg":"storing future round for later","process":"agreement","round":45146,"time":"2021-04-16T14:30:11Z","topic":"Agreement"}
{"error":"read tcp 159.89.24.209:38204-\u003e46.101.134.69:7100: i/o timeout","level":"warning","msg":"error reading message","process":"peer","time":"2021-04-16T14:29:58Z"}
{"error":"read tcp 159.89.24.209:38204-\u003e46.101.134.69:7100: i/o timeout","level":"warning","msg":"error received from peer","process":"server","time":"2021-04-16T14:30:11Z"}
{"error":"write tcp 159.89.24.209:7100-\u003e134.122.34.54:50876: write: broken pipe","level":"error","msg":"got error back from keepAlive","process":"keepaliveloop","time":"2021-04-16T14:29:58Z"}
{"error":"write tcp 159.89.24.209:55820-\u003e165.227.237.8:7100: write: broken pipe","level":"error","msg":"got error back from keepAlive","process":"keepaliveloop","time":"2021-04-16T14:29:58Z"}
{"error":"write tcp 159.89.24.209:38204-\u003e46.101.134.69:7100: write: broken pipe","level":"warning","msg":"error writing message","process":"writeloop","time":"2021-04-16T14:29:58Z"}
```

Right after these entries, the logs indicate a new startup.

--- a/cmd/dusk/server.go
+++ b/cmd/dusk/server.go
@@ -307,14 +307,7 @@ func (s *Server) OnAccept(conn net.Conn) {
 	logServer.WithField("address", peerReader.Addr()).Debugln("connection established")
 
 	peerWriter := peer.NewWriter(conn, s.gossip, s.eventBus)
-
-	go func() {
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-
-		err := peer.Create(ctx, peerReader, peerWriter, writeQueueChan)
-		logServer.WithError(err).Warnln("error received from peer")
-	}()
+	go peer.Create(context.Background(), peerReader, peerWriter, writeQueueChan)
 }
 
 // OnConnection is the callback for writing to the peers.
@@ -333,14 +326,7 @@ func (s *Server) OnConnection(conn net.Conn, addr string) {
 		Debugln("connection established")
 
 	peerReader := s.readerFactory.SpawnReader(conn, s.gossip, writeQueueChan)
-
-	go func() {
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-
-		err := peer.Create(ctx, peerReader, peerWriter, writeQueueChan)
-		logServer.WithError(err).Warnln("error received from peer")
-	}()
+	go peer.Create(context.Background(), peerReader, peerWriter, writeQueueChan)
 }
 
 // Close the chain and the connections created through the RPC bus.

--- a/pkg/p2p/peer/peer.go
+++ b/pkg/p2p/peer/peer.go
@@ -191,9 +191,9 @@ func (p *Reader) Accept() error {
 	return nil
 }
 
-// Create two-way communication with a peer. This function will allow both
-// goroutines to run as long as no errors are encountered. Once the first error
-// comes through, the context is canceled, and both goroutines are cleaned up.
+// Create two-way communication with a peer. The function creates a shared context
+// between both goroutines, which allows one to cancel the other in case of a
+// fatal error being encountered.
 func Create(ctx context.Context, reader *Reader, writer *Writer, writeQueueChan <-chan bytes.Buffer) {
 	pCtx, cancel := context.WithCancel(ctx)
 

--- a/pkg/p2p/peer/peer.go
+++ b/pkg/p2p/peer/peer.go
@@ -194,27 +194,22 @@ func (p *Reader) Accept() error {
 // Create two-way communication with a peer. This function will allow both
 // goroutines to run as long as no errors are encountered. Once the first error
 // comes through, the context is canceled, and both goroutines are cleaned up.
-func Create(ctx context.Context, reader *Reader, writer *Writer, writeQueueChan <-chan bytes.Buffer) error {
-	errChan := make(chan error, 1)
-	defer close(errChan)
+func Create(ctx context.Context, reader *Reader, writer *Writer, writeQueueChan <-chan bytes.Buffer) {
+	pCtx, cancel := context.WithCancel(ctx)
 
-	go reader.ReadLoop(ctx, errChan)
-	go writer.Serve(ctx, writeQueueChan, errChan)
+	go func() {
+		reader.ReadLoop(pCtx)
+		cancel()
+	}()
 
-	return <-errChan
-}
-
-// This attempts a non-blocking send to the errChan, which prevents panics
-// from sending to a closed channel.
-func sendError(errChan chan error, err error) {
-	select {
-	case errChan <- err:
-	default:
-	}
+	go func() {
+		writer.Serve(ctx, writeQueueChan)
+		cancel()
+	}()
 }
 
 // Serve utilizes two different methods for writing to the open connection.
-func (w *Writer) Serve(ctx context.Context, writeQueueChan <-chan bytes.Buffer, errChan chan error) {
+func (w *Writer) Serve(ctx context.Context, writeQueueChan <-chan bytes.Buffer) {
 	// Any gossip topics are written into interrupt-driven ringBuffer
 	// Single-consumer pushes messages to the socket
 	g := &GossipConnector{w.gossip, w.Connection}
@@ -222,7 +217,7 @@ func (w *Writer) Serve(ctx context.Context, writeQueueChan <-chan bytes.Buffer, 
 
 	// writeQueue - FIFO queue
 	// writeLoop pushes first-in message to the socket
-	w.writeLoop(ctx, writeQueueChan, errChan)
+	w.writeLoop(ctx, writeQueueChan)
 	w.onDisconnect()
 }
 
@@ -262,7 +257,7 @@ func (w *Writer) onDisconnect() {
 	}
 }
 
-func (w *Writer) writeLoop(ctx context.Context, writeQueueChan <-chan bytes.Buffer, errChan chan error) {
+func (w *Writer) writeLoop(ctx context.Context, writeQueueChan <-chan bytes.Buffer) {
 	defer func() {
 		if config.Get().API.Enabled {
 			go func() {
@@ -292,8 +287,6 @@ func (w *Writer) writeLoop(ctx context.Context, writeQueueChan <-chan bytes.Buff
 
 			if _, err := w.Connection.Write(buf.Bytes()); err != nil {
 				l.WithField("process", "writeloop").WithError(err).Warnln("error writing message")
-				sendError(errChan, err)
-
 				return
 			}
 		case <-ctx.Done():
@@ -306,7 +299,7 @@ func (w *Writer) writeLoop(ctx context.Context, writeQueueChan <-chan bytes.Buff
 // ReadLoop will block on the read until a message is read, or until the deadline
 // is reached. Should be called in a go-routine, after a successful handshake with
 // a peer. Eventual duplicated messages are silently discarded.
-func (p *Reader) ReadLoop(ctx context.Context, errChan chan error) {
+func (p *Reader) ReadLoop(ctx context.Context) {
 	// As the peer ReadLoop is at the front-line of P2P network, receiving a
 	// malformed frame by an adversary node could lead to a panic.
 	// In such situation, the node should survive but adversary conn gets dropped
@@ -315,10 +308,10 @@ func (p *Reader) ReadLoop(ctx context.Context, errChan chan error) {
 	// 		log.Errorf("Peer %s failed with critical issue: %v", p.RemoteAddr(), r)
 	// 	}
 	// }()
-	p.readLoop(ctx, errChan)
+	p.readLoop(ctx)
 }
 
-func (p *Reader) readLoop(ctx context.Context, errChan chan error) {
+func (p *Reader) readLoop(ctx context.Context) {
 	defer func() {
 		_ = p.Conn.Close()
 	}()
@@ -344,27 +337,23 @@ func (p *Reader) readLoop(ctx context.Context, errChan chan error) {
 		err := p.Conn.SetReadDeadline(time.Now().Add(readWriteTimeout))
 		if err != nil {
 			l.WithError(err).Warnf("error setting read timeout")
-			sendError(errChan, err)
 			return
 		}
 
 		b, err := p.gossip.ReadMessage(p.Conn)
 		if err != nil {
 			l.WithError(err).Warnln("error reading message")
-			sendError(errChan, err)
 			return
 		}
 
 		message, cs, err := checksum.Extract(b)
 		if err != nil {
 			l.WithError(err).Warnln("error reading Extract message")
-			sendError(errChan, err)
 			return
 		}
 
 		if !checksum.Verify(message, cs) {
 			l.WithError(errors.New("invalid checksum")).Warnln("error reading message")
-			sendError(errChan, err)
 			return
 		}
 

--- a/pkg/p2p/peer/peer_in_test.go
+++ b/pkg/p2p/peer/peer_in_test.go
@@ -229,7 +229,7 @@ func testReader(t *testing.T, f *ReaderFactory) (*Reader, net.Conn, net.Conn, ch
 	peer := f.SpawnReader(r, g, respChan)
 
 	// Run the non-recover readLoop to watch for panics
-	go assert.NotPanics(t, func() { peer.readLoop(context.Background(), make(chan error, 1)) })
+	go assert.NotPanics(t, func() { peer.readLoop(context.Background()) })
 
 	time.Sleep(200 * time.Millisecond)
 

--- a/pkg/p2p/peer/peer_test.go
+++ b/pkg/p2p/peer/peer_test.go
@@ -55,7 +55,7 @@ func TestReader(t *testing.T) {
 	responseChan := make(chan bytes.Buffer, 100)
 	peerReader := factory.SpawnReader(srv, protocol.NewGossip(protocol.TestNet), responseChan)
 
-	go peerReader.ReadLoop(context.Background(), make(chan error, 1))
+	go peerReader.ReadLoop(context.Background())
 
 	errChan := make(chan error, 1)
 	go func(eChan chan error) {
@@ -116,7 +116,7 @@ func TestWriteLoop(t *testing.T) {
 		responseChan := make(chan bytes.Buffer)
 
 		writer := NewWriter(client, g, bus, 30*time.Millisecond)
-		go writer.Serve(context.Background(), responseChan, make(chan error, 1))
+		go writer.Serve(context.Background(), responseChan)
 
 		bufCopy := buf
 		responseChan <- bufCopy

